gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenAuthorityService.java

```diff
@@ -232,13 +236,18 @@ public boolean verifyToken(JWT token, String jwksurl, String algorithm, Set<JOSE
         long cacheTTL = config.getJwksCacheTimeToLive();
         long cacheTimeOut = config.getJwksCacheRefreshTimeout();
 
-        JWKSource<SecurityContext> keySource = JWKSourceBuilder.create(new URL(jwksurl))
-                .cache(cacheTTL, cacheTimeOut)
-                .refreshAheadCache(true)
-                .retrying(true)
-                .outageTolerant(outageTTL)
-                .build();
-        JWSKeySelector<SecurityContext> keySelector = new JWSVerificationKeySelector<>(expectedJWSAlg, keySource);
+        if(cachedJwkSources.get(jwksurl) == null) {
+          final JWKSource<SecurityContext> keySource = JWKSourceBuilder.create(new URL(jwksurl))
+                  .cache(cacheTTL, cacheTimeOut)
+                  .refreshAheadCache(true)
+                  .retrying(true)
+                  .outageTolerant(outageTTL)
+                  .build();
+            cachedJwkSources.put(jwksurl, keySource);
+        }
+
+
+        JWSKeySelector<SecurityContext> keySelector = new JWSVerificationKeySelector<>(expectedJWSAlg, cachedJwkSources.get(jwksurl));
```

Comment:
Seems like you could do the get only once on the cachedJwkSources.

Reply:
Sure, i'll update the PR